### PR TITLE
Make loop_emscripten stateful and allow for multiple instances.

### PR DESCRIPTION
**Pull request by ScottTodd**
Follow-up to https://github.com/iree-org/iree/pull/11204.

* In C, each `iree_loop_emscripten_t` now tracks a `iree_loop_emscripten_scope_t` handle
* In JS, some new classes are added:
  * `LoopEmscriptenScope`: a collection of pending operations that can be aborted together
  * `LoopCommand` / `LoopCommandCall`: polymorphic containers for pending operations with support for aborting. The plan is to add more classes for other operations as I implement them.

I'm still testing using loop_test.h (which depends on `iree_loop_drain()` and thus falls over immediately when this yields to the event loop), but I've started work on an alternate implementation written in JS.

### CODE REVIEW

File: runtime/src/iree/base/loop_emscripten.js [hunk 1]
```diff
@@ -14,28 +14,102 @@
...
-        const IREE_STATUS_OK = 0;
+        this.callback = callback;
+        this.user_data = user_data;
+        this.loop = loop;
 
-        setTimeout(() => {
-          const ret =
-              Module['dynCall_iiii'](callback, user_data, loop, IREE_STATUS_OK);
+        this.timeoutId = setTimeout(() => {
+          getWasmTableEntry(callback)(user_data, loop, IREE_STATUS_OK);
```
Comment (ScottTodd) [inline runtime/src/iree/base/loop_emscripten.js:36]: Still not sure if this is a stable way to call function pointers. It seems to work for some small cases but I see a few errors:

```js
// These appear to be equivalent, but both sometimes return
// `RuntimeError: table index is out of bounds`
getWasmTableEntry(callback)(user_data, loop, IREE_STATUS_OK);
{{{ makeDynCall('iiii', 'callback') }}} (user_data, loop, IREE_STATUS_OK);

// This is what I had before, but it sometimes returns
// `TypeError: Module.dynCall_iiii is not a function`
Module['dynCall_iiii'](callback, user_data, loop, IREE_STATUS_OK);
```

Should probably check with the Emscripten devs.

Comment (ScottTodd) [reply runtime/src/iree/base/loop_emscripten.js]: Ehhhh, read through some PRs/Issues and settled on something that I think works... (use the generic `dynCall` with `-sDYNCALLS=1 -sEXPORTED_RUNTIME_METHODS=['dynCall']` linkopts)

Here's the latest source I'm referencing: https://github.com/emscripten-core/emscripten/blob/22b137742d40da3949a0e97cfc8e8172e41e2ace/test/test_core.py#L7098-L7120

Comment (benvanik) [reply runtime/src/iree/base/loop_emscripten.js]: I think you need to declare it as a dep:
https://github.com/emscripten-core/emscripten/blob/1f0e614acf895d6b75d693227b99cd3c89a4f30e/ChangeLog.md?plain=1#L267-L269
so add to `DEFAULT_LIBRARY_FUNCS_TO_INCLUDE` or `EXPORTED_RUNTIME_METHODS` or better specify it on the method:
https://github.com/emscripten-core/emscripten/blob/1f0e614acf895d6b75d693227b99cd3c89a4f30e/src/library_addfunction.js#L165

File: runtime/src/iree/base/loop_emscripten.c [hunk 2]
```diff
@@ -16,7 +16,13 @@
 // externs from loop_emscripten.js
 //===----------------------------------------------------------------------===//
 
-extern iree_status_t loop_command_call(iree_loop_callback_fn_t callback,
+typedef uint32_t iree_loop_emscripten_scope_t;  // Opaque handle.
+
+extern iree_loop_emscripten_scope_t loop_allocate_scope();
+extern void loop_free_scope(iree_loop_emscripten_scope_t scope);
```
Comment (benvanik) [inline runtime/src/iree/base/loop_emscripten.c:22]: prefix with iree_ (and loop_command_call too)

File: runtime/src/iree/base/loop_emscripten.js [hunk 3]
```diff
@@ -14,28 +14,104 @@
...
+    const IREE_STATUS_CODE_MASK = 0x1F;
+    const IREE_STATUS_ABORTED = 10 & IREE_STATUS_CODE_MASK;
+    const IREE_STATUS_OUT_OF_RANGE = 11 & IREE_STATUS_CODE_MASK;
+
+    class LoopOperation {
+      abort() {}
+    }
+
+    // IREE_LOOP_COMMAND_CALL
+    class LoopOperationCall extends LoopOperation {
```
Comment (benvanik) [inline runtime/src/iree/base/loop_emscripten.js:27]: nit: LoopCommand / LoopCommandCall to match the C IREE_LOOP_COMMAND_CALL/etc

File: runtime/src/iree/base/loop_emscripten.js [hunk 4]
```diff
@@ -14,28 +14,104 @@
...
+        this.user_data = user_data;
+        this.loop = loop;
 
-        setTimeout(() => {
-          const ret =
-              Module['dynCall_iiii'](callback, user_data, loop, IREE_STATUS_OK);
+        this.timeoutId = setTimeout(() => {
+          Module['dynCall'](
+              'iiii', this.callback, this.user_data, this.loop, IREE_STATUS_OK);
           // TODO(scotttodd): handle the returned status (sticky failure state?)
```
Comment (benvanik) [inline runtime/src/iree/base/loop_emscripten.js:38]: note that the status will leak here if dropped

**Review by benvanik — CHANGES_REQUESTED** [review]
lgtm besides minor nits and using getWasmTableEntry

File: runtime/src/iree/base/loop_emscripten.js (hunk 1 above)
Comment (ScottTodd) [reply runtime/src/iree/base/loop_emscripten.js]: The CallFork test case is pretty unique in that it
1) In C++ (wasm) defines a callback function using a closure
2) In JS, that callback function is run within `setTimeout()`
3) The callback function itself defines a new callback function using a closure
4) In JS, the second callback function is run within `setTimeout()`

The other test cases don't have steps 3-4.

I'm thinking that somehow the `Module['dynCall'](` form is able to handle that reentrancy while `{{{ makeDynCall }}}` / `dynCall_iiii.apply` is not.

Source: https://github.com/iree-org/iree/blob/33ace455f664e93654cd60646d106a35a195a54c/runtime/src/iree/base/loop_test.h#L67-L95

**Review by benvanik — APPROVED** [review]

**Pull request merged**